src/Microsoft.AspNet.Mvc.Core/CorsAuthorizationFilter.cs

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Cors.Core;
+using Microsoft.AspNet.Http;
+using Microsoft.AspNet.WebUtilities;
+using Microsoft.Framework.DependencyInjection;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.AspNet.Mvc
+{
+    /// <summary>
+    /// A filter which applies the given <see cref="CorsPolicy"/> and adds appropriate response headers.
+    /// </summary>
+    public class CorsAuthorizationFilter : ICorsAuthorizationFilter
+    {
+        private readonly CorsPolicy _corsPolicy;
+
+        /// <summary>
+        /// Creates a new instace of <see cref="CorsAuthorizationFilter"/>.
+        /// </summary>
+        /// <param name="corsPolicy">The <see cref="CorsPolicy"/> which needs to be applied.</param>
+        public CorsAuthorizationFilter([NotNull] CorsPolicy corsPolicy)
+        {
+            _corsPolicy = corsPolicy;
+        }
+
+        /// <inheritdoc />
+        public async Task OnAuthorizationAsync([NotNull] AuthorizationContext context)
+        {
+            // If this filter is not closest to the action, it is not applicable.
+            if (!IsClosestToAction(context.ActionDescriptor))
+            {
+                return;
+            }
+
+            if (context.HttpContext.Request.Headers.ContainsKey(CorsConstants.Origin))
```

Comment:
lots of dotting to access the request and the headers. some locals would make this much more readable.